pac4j-saml-config-db/src/main/java/org/pac4j/saml/dbclient/DbLoadedSamlClients.java

```diff
@@ -0,0 +1,197 @@
+import org.pac4j.core.client.Clients;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.saml.client.SAML2Client;
+import org.pac4j.saml.dbclient.dao.api.SamlClientDao;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An extension to base {@link Clients}, which works also with clients whose definition is stored in a database. 
+ *
+ * @author jkacer
+ * @since 1.9.0
+ */
+public class DbLoadedSamlClients extends Clients {
+
+    /** SLF4J logger. */
+    private final Logger logger = LoggerFactory.getLogger(DbLoadedSamlClients.class);
+    
+    /** DAO reading definitions of SAML clients from a database. */
+    private SamlClientDao samlClientDao;
+
+    /** Clients set explicitly using a setter. */
+    private List<Client> explicitClients;
+    
+    // ------------------------------------------------------------------------------------------------------------------------------------
+
+    
+	/**
+	 * Creates a new client group with no explicit clients.
+	 */
+	public DbLoadedSamlClients() {
+		super();
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients.
+	 * 
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(Client... clients) {
+		super(clients);
+	}
+
+
+	/**
+	 * Creates a new client group with a single explicit client.
+	 * 
+	 * @param client
+	 *            The client.
+	 */
+	public DbLoadedSamlClients(Client client) {
+		super(client);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients.
+	 * 
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(List<Client> clients) {
+		super(clients);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 * 
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, Client... clients) {
+		super(callbackUrl, clients);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 * 
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param client
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, Client client) {
+		super(callbackUrl, client);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 *
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, List<Client> clients) {
+		super(callbackUrl, clients);
+	}
+
+
+	/**
+	 * Initializes all clients by loading them from the database.
+	 * 
+	 * @see org.pac4j.core.util.InitializableWebObject#internalInit(org.pac4j.core.context.WebContext)
+	 */
+	@Override
+	protected void internalInit(final WebContext context) {
+        // Remove all clients first if re-initialized

```

Comment:
Hi! It should be done now. The AdditionalClientContributor is gone, there is a new specialized clients class that subclasses Clients, overrides getClients(); there is one additional internal method.